backend/cellguide/pipeline/ontology_tree/tree_builder.py

```diff
@@ -205,3 +213,332 @@ def _traverse_ontology_with_counting(
+                self._truncate_graph_first_pass(ontology_graph_copy, valid_nodes)
+                self._truncate_graph_second_pass(ontology_graph_copy, visited_nodes_in_paths)
+
+                isExpandedNodes = list(set(_getExpandedData(ontology_graph_copy)))
+                notShownWhenExpandedNodes = _getShownData(ontology_graph_copy)
+
+                notShownWhenExpanded = {}
+                for i in notShownWhenExpandedNodes:
+                    notShownWhenExpanded.update(i)
+                all_states_per_cell_type[end_node] = {
+                    "isExpandedNodes": isExpandedNodes,
+                    "notShownWhenExpandedNodes": notShownWhenExpanded,
+                }
+
+        return all_states_per_cell_type
+
+    def _depth_first_search_pathfinder(self, path_end_node, node=None, path=None):
+        """
+        This function finds a path backwards from the end node to the start node using depth-first search.
+
+        Arguments
+        ---------
+        path_end_node: str
+            The end node of the path.
+        node: str, optional, default=None
+            The current node in the depth-first search.
+        path: list, optional, default=None
+            The current path in the depth-first search.
+
+        Returns
+        -------
+        path: list
+            The path from the end node to the start node.
+        """
+        if path is None:
+            path = [path_end_node]
+            node = path_end_node
+
+        if node == self.start_node:
+            return path
+
+        for parent in self.all_parents.get(node, []):
+            full_path = self._depth_first_search_pathfinder(path_end_node, node=parent, path=path + [parent])
+            if full_path:
+                return full_path
+
+    def _truncate_graph_first_pass(self, graph, valid_nodes):
+        """
+        This function truncates the ontology graph in-place by removing nodes that are not in the valid nodes list.
+        It also adds a dummy node to the graph if the graph has children that are not in the valid nodes list.
+        The dummy node contains the ids of the invalid children.
+
+        Arguments
+        ---------
+        graph: dict
+            A node of the ontology graph. At the top-level, this should be the root node of the ontology graph.
+        valid_nodes: list
+            A list of valid cell type ontology term ids.
+        """
+        if graph["id"] not in valid_nodes:
+            return False
+
+        children = graph.get("children", [])
+        valid_children = []
+        append_dummy = False
+
+        invalid_children_ids = []
+        for child in children:
+            is_valid = self._truncate_graph_first_pass(child, valid_nodes)
+            if is_valid:
+                valid_children.append(child)
+            elif child["id"] != "":
+                invalid_children_ids.append(child["id"])
+                append_dummy = True
+
+        if append_dummy and len(valid_children) > 0:
```

Comment:
i think you can remove `append_dummy` and just write `if len(invalid_children_ids) > 0 and len(valid_children) > 0:` which also makes it a bit clearer what the case is for appending a dummy node. you need to have a mix of invalid and valid children. only invalid children should result in removing the children rather than appending a dummy node